diff --git a/Cargo.toml b/Cargo.toml
index 40ac0ce..046303a 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -7,7 +7,7 @@ rust-version = "1.88"
 
 [lib]
 name = "ssl"
-crate-type = ["cdylib"]
+crate-type = ["staticlib"]
 
 [dependencies]
 env_logger = "0.11"
diff --git a/build.rs b/build.rs
index ff8b8f6..642af48 100644
--- a/build.rs
+++ b/build.rs
@@ -1,22 +1,6 @@
 use std::{env, fs, path};
 
 fn main() {
-    if cfg!(target_os = "linux") {
-        println!("cargo:rustc-cdylib-link-arg=-Wl,--soname=libssl.so.3");
-
-        // We require lld, because ld only supports one --version-script
-        // and rustc uses it for its own purposes (and provides no API for us).
-        println!("cargo:rustc-cdylib-link-arg=-fuse-ld=lld");
-
-        let filename = write_version_file();
-        println!("cargo:rustc-cdylib-link-arg=-Wl,--version-script={filename}");
-
-        for symbol in ENTRYPOINTS {
-            // Rename underscore-prefixed symbols (produced by rust code) to
-            // unprefixed symbols (manipulated by our version file).
-            println!("cargo:rustc-cdylib-link-arg=-Wl,--defsym={symbol}=_{symbol}",);
-        }
-    }
 }
 
 fn write_version_file() -> String {
